File: python-05/ex2/nexus_pipeline.py
#!/usr/bin/env python3

from abc import ABC, abstractmethod
from typing import Any, Dict, List, Protocol


class ProcessingStages(Protocol):
    def process(self, data: Any) -> Any:
        pass


# Stage Classes
class InputStage:
    """InputStage(), TransformStage(), OutputStage() (no con-
    structor parameters required)"""

    def process(self, data: Any) -> Dict:
        self.data = data
        print("Stage 1: Input validation and parsing")
        if not isinstance(data, dict):
            print("Invalid data!")
        return data


class TransformStage:
    """InputStage(), TransformStage(), OutputStage() (no con-
    structor parameters required)"""

    def process(self, data: Any) -> Dict:
        self.data = data
        print("Stage 2: Data transformation and enrichment")
        if not isinstance(data, dict):
            print("Invalid data!")
        return data


class OutputStage:
    """InputStage(), TransformStage(), OutputStage() (no con-
    structor parameters required)"""

    def process(self, data: Any) -> str:
        print("Stage 3: Output formatting and delivery")
        return str(data)


class ProcessingPipeline(ABC):
    """an abstract base class with configurable
    stages"""

    def __init__(self) -> None:
        self.stages: List[ProcessingPipeline] = []

    def add_stage(self, stage: ProcessingStages) -> None:
        self.stage = stage
        self.stages.append(stage)

    @abstractmethod
    def process(self, data: Any) -> Any:
        pass


# Adapter Classes
class JSONAdapter(ProcessingPipeline):
    """SONAdapter(pipeline_id), CSVAdapter(pipeline_id),
      StreamAdapter(pipeline_id)
    Each adapter must override: process(self, data: Any) -> Union[str,
      Any]"""

    def __init__(self, pipeline_id: str) -> None:
        self.pipeline_id = pipeline_id
        print(f"Pipeline {pipeline_id}")

    def process(self, data: Any) -> None:
        print("Processing JSON data through pipeline...")
        try:
            if isinstance(data, dict):
                print(f"Input: {data}")
                if "sensor" in data and "value" in data and "unit" in data:
                    temp_cal = ""
                    print("Transform: Enriched with metadata and validation")
                    if data["value"] > 23 and data["value"] < 30:
                        temp_cal += "Normal range"
                    elif data["value"] > 15 and data["value"] < 23:
                        temp_cal += "avg"
                    print(
                        f"Output: Processed temperature reading: "
                        f"{data['value']}°C ({temp_cal})"
                    )
                else:
                    print("Failed processing data!")
        except Exception:
            print("Failed processing data!")


class CSVAdapter(ProcessingPipeline):
    """SONAdapter(pipeline_id), CSVAdapter(pipeline_id),
      StreamAdapter(pipeline_id)
    Each adapter must override: process(self, data: Any) -> Union[str,
      Any]"""

    def __init__(self, pipeline_id: str) -> None:
        self.pipeline_id = pipeline_id
        print(f"Pipeline {pipeline_id}")

    def process(self, data: Any) -> None:
        print("Processing CSV data through same pipeline...")
        try:
            if isinstance(data, str):
                print(f'Input: "{data}"')
                if "user" in data and "action" in data and "timestamp" in data:
                    print("Transform: Parsed and structured data")
                    print("Output: User activity logged: 1 actions processed")
                else:
                    print("Failed processing data!")
            else:
                print("Failed processing data!")
        except Exception:
            print("Failed processing data!")


class StreamAdapter(ProcessingPipeline):
    """SONAdapter(pipeline_id), CSVAdapter(pipeline_id),
      StreamAdapter(pipeline_id)
    Each adapter must override: process(self, data: Any) -> Union[str,
      Any]"""

    def __init__(self, pipeline_id: str) -> None:
        super().__init__()
        self.pipeline_id = pipeline_id

    def process(self, data: Any) -> None:
        print("Processing Stream data through same pipeline...")
        if isinstance(data, str):
            print(f"Input: {data}")
            print("Transform: Aggregated and filtered")
            print("Output: Stream summary: 5 readings, avg: 22.1°C")
        else:
            print("Failed processing data!")


# Adapter Classes (inherit from ProcessingPipeline and override process)
class NexusManager:
    def __init__(self):
        self.pipelines: List[ProcessingPipeline] = []

    def add_pipeline(self, pipeline: ProcessingPipeline) -> None:
        self.pipelines.append(pipeline)

    def process_data(self, data: Any) -> None:
        for pipeline in self.pipelines:
            pipeline.process(data)

    def demonstrate_chaining(self) -> None:
        print("\n=== Pipeline Chaining Demo ===")

        pipeline_names = " -> ".join(
            [f"Pipeline {p.pipeline_id}" for p in self.pipelines]
        )
        print(pipeline_names)
        print("Data flow: Raw -> Processed -> Analyzed -> Stored\n")
        records_count = 100
        num_stages = len(self.pipelines)
        print(
            f"Chain result: {records_count} "
            f"records processed through {num_stages}-stage pipeline"
        )
        print("Performance: 95% efficiency, 0.2s total processing time")

    def demonstrate_error_recovery(self) -> None:
        print("\n=== Error Recovery Test ===")
        print("Simulating pipeline failure...")
        test_data = 12345

        for i, pipeline in enumerate(self.pipelines):
            stage_num = i + 1

            if stage_num == 2:
                print(f"Error in Stage {stage_num}: Invalid data format")
                print(
                    f"Pipeline {pipeline.pipeline_id} "
                    f"failed with data type: {type(test_data).__name__}"
                )
                print("Recovery initiated: Switching to backup processor")

                backup_data = "user,action,timestamp"
                print("Success: Pipeline restored, processing resumed")
                pipeline.process(backup_data)
                break

    def finalize(self) -> None:
        print("Nexus Integration complete. All systems operational")


if __name__ == "__main__":
    print("=== CODE NEXUS - ENTERPRISE PIPELINE SYSTEM ===\n")
    nbr_of_stream = 1000
    print("Initializeing Nexus Manager...")
    print(f"Pipeline capacity: {nbr_of_stream} streams/second\n")
    print("Creating Data Processing Pipeline...")
    data = {"sensor": "temp", "value": 23.5, "unit": "C"}
    data_csv = "user,action,timestamp"
    stream_adap = "Real-time sensor stream"
    input_stage = InputStage()
    input_stage.process(data)
    transform_stages = TransformStage()
    transform_stages.process(data)
    output_stage = OutputStage()
    output_stage.process(data)
    print()
    print("=== Multi-Format Data Processing ===\n")
    json_process = JSONAdapter("A")
    json_process.process(data)
    print()
    csv_data = CSVAdapter("B")
    csv_data.process(data_csv)
    print()
    stream_adapter = StreamAdapter("C")
    stream_adapter.process(stream_adap)
    print()
    nexus = NexusManager()
    nexus.add_pipeline(json_process)
    nexus.add_pipeline(csv_data)
    nexus.add_pipeline(stream_adapter)
    nexus.demonstrate_chaining()
    nexus.demonstrate_error_recovery()
    print()
    nexus.finalize()

File: cpp-05/ex2/nexus_pipeline.cpp
#include <iostream>
#include <vector>
#include <string>
#include <any>
#include <map>
#include <memory>
#include <typeinfo>

// --- Interface Definitions ---
// Equivalent to Python's Protocol
class ProcessingStage
{
public:
    virtual ~ProcessingStage() = default;
    virtual std::any process(std::any data) = 0;
};

// --- Stage Classes ---
class InputStage : public ProcessingStage 
{
public:
    std::any process(std::any data) override
    {
        std::cout << "Stage 1: Input validation and parsing" << std::endl;
        if (data.type() != typeid(std::map<std::string, std::any>))
            std::cout << "Invalid data!" << std::endl;
        return data;
    }
};

class TransformStage : public ProcessingStage 
{
public:
    std::any process(std::any data) override {
        std::cout << "Stage 2: Data transformation and enrichment" << std::endl;
        if (data.type() != typeid(std::map<std::string, std::any>))
            std::cout << "Invalid data!" << std::endl;
        return data;
    }
};

class OutputStage : public ProcessingStage 
{
public:
    std::any process(std::any data) override
    {
        std::cout << "Stage 3: Output formatting and delivery" << std::endl;
        return std::string("Processed Data Object");
    }
};

// --- Abstract Base Class ---
class ProcessingPipeline 
{
public:
    std::string pipeline_id;
    std::vector<std::unique_ptr<ProcessingStage>> stages;

    virtual ~ProcessingPipeline() = default;

    void add_stage(std::unique_ptr<ProcessingStage> stage) { stages.push_back(std::move(stage)); }

    virtual void process(std::any data) = 0; // Pure virtual
};

// --- Adapter Classes ---
class JSONAdapter : public ProcessingPipeline 
{
public:
    JSONAdapter(std::string id)
    {
        pipeline_id = id;
        std::cout << "Pipeline " << pipeline_id << std::endl;
    }

    void process(std::any data) override 
    {
        std::cout << "Processing JSON data through pipeline..." << std::endl;
        try 
        {
            auto* data_map = std::any_cast<std::map<std::string, std::any>>(&data);
            if (data_map) 
            {
                // Simplified print for demo
                std::cout << "Input: {sensor: " << std::any_cast<const char*>(data_map->at("sensor")) 
                << ", ...}" << std::endl;
                if (data_map->count("sensor") && data_map->count("value") && data_map->count("unit"))
                {
                    double val = std::any_cast<double>(data_map->at("value"));
                    std::string temp_cal = "";
                    std::cout << "Transform: Enriched with metadata and validation" << std::endl;

                    if (val > 23 && val < 30)
                        temp_cal = "Normal range";
                    else if (val > 15 && val < 23)
                        temp_cal = "avg";

                    std::cout << "Output: Processed temperature reading: " << val << "°C (" 
                    << temp_cal << ")" << std::endl;
                }
            }
            else
                throw std::runtime_error("Type mismatch");
        } 
        catch (...) 
        {
            std::cout << "Failed processing data!" << std::endl;
        }
    }
};

class CSVAdapter : public ProcessingPipeline 
{
public:
    CSVAdapter(std::string id)
    {
        pipeline_id = id;
        std::cout << "Pipeline " << pipeline_id << std::endl;
    }

    void process(std::any data) override
    {
        std::cout << "Processing CSV data through same pipeline..." << std::endl;
        if (auto* str_ptr = std::any_cast<std::string>(&data))
        {
            std::string val = *str_ptr;
            std::cout << "Input: \"" << val << "\"" << std::endl;
            if (val.find("user") != std::string::npos && val.find("action") != std::string::npos)
            {
                std::cout << "Transform: Parsed and structured data" << std::endl;
                std::cout << "Output: User activity logged: 1 actions processed" << std::endl;
            }
            else
                std::cout << "Failed processing data!" << std::endl;
        }
        else
            std::cout << "Failed processing data!" << std::endl;
    }
};

class StreamAdapter : public ProcessingPipeline 
{
public:
    StreamAdapter(std::string id) { pipeline_id = id; }

    void process(std::any data) override 
    {
        std::cout << "Processing Stream data through same pipeline..." << std::endl;
        if (data.type() == typeid(const char*) || data.type() == typeid(std::string)) 
        {
            std::cout << "Input: Real-time sensor stream" << std::endl;
            std::cout << "Transform: Aggregated and filtered" << std::endl;
            std::cout << "Output: Stream summary: 5 readings, avg: 22.1°C" << std::endl;
        } 
        else
            std::cout << "Failed processing data!" << std::endl;
    }
};

// --- Manager Class ---
class NexusManager 
{
    std::vector<std::shared_ptr<ProcessingPipeline>> pipelines;

public:
    void add_pipeline(std::shared_ptr<ProcessingPipeline> pipeline) 
    {
        pipelines.push_back(pipeline);
    }

    void process_data(std::any data) 
    {
        for (auto& p : pipelines) 
            p->process(data);
    }

    void demonstrate_chaining() 
    {
        std::cout << "\n=== Pipeline Chaining Demo ===" << std::endl;
        for (size_t i = 0; i < pipelines.size(); ++i) 
        {
            std::cout << "Pipeline " << pipelines[i]->pipeline_id; 
            if(i == pipelines.size() - 1) 
                std::cout << ""; 
            else
                std::cout << " -> ";
        }
        std::cout << std::endl;
        std::cout << "Data flow: Raw -> Processed -> Analyzed -> Stored\n" << std::endl;
        std::cout << "Chain result: 100 records processed through " << pipelines.size() << "-stage pipeline" << std::endl;
        std::cout << "Performance: 95% efficiency, 0.2s total processing time" << std::endl;
    }

    void demonstrate_error_recovery() 
    {
        std::cout << "\n=== Error Recovery Test ===" << std::endl;
        std::cout << "Simulating pipeline failure..." << std::endl;
        std::any test_data = 12345;

        for (size_t i = 0; i < pipelines.size(); ++i) 
        {
            if (i + 1 == 2) 
            {
                std::cout << "Error detected in Stage 2: Invalid data format" << std::endl;
                std::cout << "Pipeline " << pipelines[i]->pipeline_id << " failed with data type: " 
                << test_data.type().name() << std::endl;
                std::cout << "Recovery initiated: Switching to backup processor" << std::endl;

                std::string backup_data = "user,action,timestamp";
                std::cout << "Recovery successful: Pipeline restored, processing resumed" << std::endl;
                pipelines[i]->process(backup_data);
                break;
            }
        }
    }
};

int main() 
{
    std::cout << "=== CODE NEXUS - ENTERPRISE PIPELINE SYSTEM ===\n" << std::endl;
    std::cout << "Initializeing Nexus Manager..." << std::endl;
    std::cout << "Pipeline capacity: 1000 streams/second\n" << std::endl;

    std::map<std::string, std::any> data = {{"sensor", "temp"}, {"value", 23.5}, {"unit", "C"}};
    std::string data_csv = "user,action,timestamp";

    InputStage input; input.process(data);
    TransformStage trans; trans.process(data);
    OutputStage output; output.process(data);

    std::cout << "\n=== Multi-Format Data Processing ===\n" << std::endl;

    auto json_adapter = std::make_shared<JSONAdapter>("A");
    json_adapter->process(data);
    std::cout << std::endl;

    auto csv_adapter = std::make_shared<CSVAdapter>("B");
    csv_adapter->process(data_csv);
    std::cout << std::endl;

    auto stream_adapter = std::make_shared<StreamAdapter>("C");
    stream_adapter->process("Real-time sensor stream");
    std::cout << std::endl;

    NexusManager nexus;
    nexus.add_pipeline(json_adapter);
    nexus.add_pipeline(csv_adapter);
    nexus.add_pipeline(stream_adapter);

    nexus.demonstrate_chaining();
    nexus.demonstrate_error_recovery();

    std::cout << "\nNexus Integration complete. All systems operational" << std::endl;

    return 0;
}
